### ISSUE TITLE
Add text/shape/other overlays

### ISSUE THREAD

**Issue by thompson318** [UCL/scikit-surgeryvtk#8]
In GitLab by @ThomasDowrick on Nov 21, 2018, 16:37

While VTK is needed for 3D model rendering, both VTK and OpenCV could be used for overlaying shapes, text, drawing stuff on the screen etc. In most cases, doing an overlay in OpenCV is somewhat simpler, and faster (the VTK pipeline requires a deep copy of the data) than doing the equivalent overlay in VTK, but VTK offers more felxibility.

a) Do everthing overlay related in VTK even if in some cases it would be easier to do the overlay in OpenCV. This would mean everything stays consistent and there are fewer dependencies for scikit-surgeryoverlay.

b) Add OpenCV as a dependency to scikit-surgeryoverlay, and do some things in OpenCV and some in VTK. In practice, I can't see someone using scikit-surgeryoverlay without also using scikit-surgeryimage, so having the extra dependency on OpenCV isn't too big a deal.

c) Something else

Thoughts? @StephenThompson @MattClarkson

**Comment by thompson318** [comment]
In GitLab by @ThomasDowrick on Nov 29, 2018, 13:31

changed title from **Add text{- overlay-}** to **Add text{+/shape/other overlays+}**

**Comment by thompson318** [comment]
In GitLab by @ThomasDowrick on Nov 29, 2018, 13:31

changed the description

**Comment by thompson318** [comment]
In GitLab by @MattClarkson on Nov 29, 2018, 13:41

When you say "doing an overlay in OpenCV is simpler", are you talking about adding it to the window that pops up when doing imshow? If so, thats a different type of window to the VTK one. So, how does an OpenCV work in a VTK window? The reason I ask, is that the other option is just to use some Qt stuff.

**Comment by thompson318** [comment]
In GitLab by @StephenThompson on Nov 29, 2018, 13:42

I'd be inclined towards doing everything in VTK. I think when you say overlay library people assume it will support 3D model rendering, so VTK has to be there, lets keep as simple as possible.

**Comment by thompson318** [comment]
In GitLab by @ThomasDowrick on Nov 29, 2018, 14:18

@MattClarkson  In OpenCV, overlaying can be done by 'drawing' directly onto the captured frame e.g.:
```
cv2.Circle(frame, center, radius, color) 

or

cv2.putText(frame, text, position, fontFace, fontScale, color)
```

The frame data can then be handled as the original frame would have been, which could be passing it to another filter, or dislpaying using vtk/imshow, or writing to disk.

The other thing that would be easier to do using OpenCV is saving a copy of the edited frame(s) as everything is already a numpy array. In VTK, it is necessary to convert the scene to a numpy array before writing, which can be a bit fiddly.

Having said this, I probably lean slightly towards doing everything in VTK, but there will be a perfomance cost to doing this and I thought it was worth a disucssion before starting any coding.

**Comment by thompson318** [comment]
In GitLab by @ThomasDowrick on Dec 3, 2018, 09:36

Decision as of 9/11/18 was to do everything in VTK where possible, so I will make a start on that.

**Comment by thompson318** [comment]
In GitLab by @ThomasDowrick on Dec 3, 2018, 09:37

created branch [`8-add-text-shape-other-overlays`](https://weisslab.cs.ucl.ac.uk/WEISS/SoftwareRepositories/SNAPPY/scikit-surgeryoverlay/compare/master...8-add-text-shape-other-overlays)

**Comment by thompson318** [comment]
In GitLab by @MattClarkson on Dec 3, 2018, 09:50

Can you ensure that decision is logged?

Also, what design are we thinking? In the longer term, I can imagine a pipeline for:

* Surfaces - in progress/done
* PointSet - another ticket once we have the main package going
* Informative text annotation at top left, top right, bottom left, bottom right.
* Emergency text annotation - REALLY BIG, scaled text, across whole of screen. Like "NO TRACKING".

If we try to make too many features, it gets unweildy, so I'd imagine just the facility to set some text, so that the end-user (application developer) can simply say create an object, and assign it to one of the 4 corners, then add to rendering pipeline.

Does that sound OK?

**Comment by thompson318** [comment]
In GitLab by @ThomasDowrick on Dec 3, 2018, 10:12

I've added the decision to the Wiki.

The above list looks good. I would also add the ability to draw a shape on the screen e.g. around an area of interest, as I have spoken to a few people who would like this functionality.

**Comment by thompson318** [comment]
In GitLab by @MattClarkson on Dec 3, 2018, 10:14

Would that be a different use-case? i.e. If there was "live" visualisation, we would do that in VTK to take advantage of OpenGL rendering. 

But drawing is always slow, so could be pure python using GUI built on Matplot lib for example? (Im just offering an alternative, not necessarily recommending it).

**Comment by thompson318** [comment]
In GitLab by @ThomasDowrick on Dec 3, 2018, 10:16

I was thinking of the case where someone would want to draw around an object in a live video (or pause it and draw round it, if it is from a file).

**Comment by thompson318** [comment]
In GitLab by @MattClarkson on Dec 3, 2018, 10:19

OK, but from a design perspective, Im just highlighting that its quite different. I'd say, just keep it simple. Do one pipeline for text, concentrating on font, text size, and positioning.

Then anything to do with drawing would be a different ticket. 

i.e. keep tickets small as poss, and simple to deliver.

**Comment by thompson318** [comment]
In GitLab by @ThomasDowrick on Dec 3, 2018, 14:19

mentioned in commit eb5b16a117741def14eeb0fc4b4cd192e0b49218

**Comment by thompson318** [comment]
In GitLab by @ThomasDowrick on Dec 3, 2018, 15:05

mentioned in commit 66f4fa3b63db081a136ff3b4f0bc44ea74033164

**Comment by thompson318** [comment]
In GitLab by @ThomasDowrick on Dec 3, 2018, 17:56

mentioned in commit 3920a2fcf040123a6acfba228d7ba5ae7a73019e

**Comment by thompson318** [comment]
In GitLab by @ThomasDowrick on Dec 3, 2018, 17:56

mentioned in commit c44febc3cf7f152c086299a2950ebf771359b08d

**Comment by thompson318** [comment]
In GitLab by @ThomasDowrick on Dec 3, 2018, 17:56

mentioned in commit 1e23c43c49d42e9021c258b3f864b38fe6179814

**Comment by thompson318** [comment]
In GitLab by @ThomasDowrick on Dec 3, 2018, 17:56

mentioned in commit 50e9daf9fbbefc50fddde9972123515b1445c2e6

**Comment by thompson318** [comment]
In GitLab by @ThomasDowrick on Dec 3, 2018, 17:58

I'll close this issue and create new issues for each type of overlay.

**Comment by thompson318** [comment]
In GitLab by @ThomasDowrick on Dec 3, 2018, 17:58

closed

**Comment by thompson318** [comment]
In GitLab by @ThomasDowrick on Dec 3, 2018, 18:02

See https://weisslab.cs.ucl.ac.uk/WEISS/SoftwareRepositories/SNAPPY/scikit-surgeryoverlay/issues/10 for text overlay

**Comment by thompson318** [comment]
In GitLab by @ThomasDowrick on Dec 4, 2018, 10:51

mentioned in commit cdf01bdd5fa8482672dcd8d89047d725e6c50e67
